composer/utils/fx_utils.py

```diff
@@ -123,16 +149,106 @@ def detect_residual_pattern(gm: GraphModule):
     raise NotImplementedError('detect_residual_pattern is currently not implemented.')
 
 
-def replace_residual_with_stochastic(gm: GraphModule):
-    """Replaces residual pattern with their stoachstic equivalent.
+def _get_ancestors(node: Node) -> List[Node]:
+    ancestorNodes = []
+    while node.op != 'placeholder':
+        ancestorNodes.append(node)
+        node = node.all_input_nodes[0]
+    return ancestorNodes
+
+
+def _get_residual_path(nodeLHS: Node, nodeRHS: Node) -> Tuple[List[Node], List[Node]]:
+    """Walk backwards from nodeLHS and nodeRSH to the root and construct lists of their parents.
 
     Arguments:
-        gm (GraphModule): The source FX-traced graph.
+        nodeLHS (Node): left-hand side node for a binary operator
+        nodeRHS (Node): right-hand side node for a binary operator
 
     Returns:
-        GraphModule: Modified GraphModule.
+        (lhsAncestors, rhsAncesotors): Two lists of nodes containing ancestors for ``nodeLHS`` and ``nodeRHS`` with
+            their common ancestors removed.
     """
-    raise NotImplementedError('replace_residual_with_stochastic is currently not implemented.')
+    lhsAncestors = _get_ancestors(nodeLHS)
+    rhsAncesotors = _get_ancestors(nodeRHS)
+
+    # Iterate from back and eliminate common nodes
+    while lhsAncestors and rhsAncesotors and lhsAncestors[-1] == rhsAncesotors[-1]:  # type: ignore
+        del lhsAncestors[-1]
+        del rhsAncesotors[-1]
+    lhsAncestors.reverse()
+    rhsAncesotors.reverse()
+    return lhsAncestors, rhsAncesotors
+
+
+def _tag_nodes(nodes: List[Node], tag: str):
+    """Tag nodes for the splitter."""
+    for node in nodes:
+        node.tag = tag  # type: ignore[attr-defined]
+
+
+def apply_stochastic_residual(gm: GraphModule, drop_rate: float = 0.2) -> Tuple[GraphModule, int]:
+    """Detect and replace residual pattern with their stochastic equivalent.
+
+    Arguments:
+        gm (GraphModule): The source FX-traced graph. It can be the whole model symbolically traced.
+
+    Returns:
+        GraphModule: Modified GraphModule that has stochastic residual connections.
+    """
+    assert isinstance(gm, GraphModule), 'Input to apply_stochastic_residual should be an instance of GraphModule'
+    count = 0
+    # all nodes that are not a part of the residual blocks are tagged with "main_0".
+    # a tag is required for all nodes by split_by_tags
+    # Also an earlier tag can be repeated for later nodes.
+    all_tags = []
+    # In this first pass over all nodes, we just tag them
+    for node in gm.graph.nodes:
+        default_tag = f'mainN_{count}'
+        node.tag = default_tag
```

Comment:
Curious if `.tag` is a part of the torchFX node spec? If not, is it possible to use a dictionary instead -- e.g. `Dict[Node, Tag]` instead of appending it onto the object?

Reply:
It is part of the torch.fx pass. See https://github.com/pytorch/pytorch/blob/master/torch/fx/passes/split_utils.py#L65